packages/js/src/generators/typescript-sync/typescript-sync.ts

```diff
@@ -0,0 +1,315 @@
+  }
+}
+
+export default syncGenerator;
+
+function updateTsConfigReferences(
+  tree: Tree,
+  tsConfigPath: string,
+  dependencies: ProjectGraphProjectNode[],
+  projectRoot: string,
+  runtimeTsConfigFileName?: string,
+  possibleRuntimeTsConfigFileNames?: string[]
+): boolean {
+  const tsConfig = readJson<Tsconfig>(tree, tsConfigPath);
+  // We have at least one dependency so we can safely set it to an empty array if not already set
+  const references = [];
+  const originalReferencesSet = new Set();
+  const newReferencesSet = new Set();
+  for (const ref of tsConfig.references ?? []) {
+    const normalizedPath = normalizeReferencePath(ref.path);
+    originalReferencesSet.add(normalizedPath);
+    if (!ref.path.startsWith('../')) {
+      // we only keep the internal references, the external ones are dictated by the project dependencies
+      references.push(ref);
+      newReferencesSet.add(normalizeReferencePath(ref.path));
+    }
+  }
+
+  let hasChanges = false;
+  for (const dep of dependencies) {
+    // Ensure the project reference for the target is set
+    let referencePath = dep.data.root;
+    if (runtimeTsConfigFileName) {
+      const runtimeTsConfigPath = joinPathFragments(
+        dep.data.root,
+        runtimeTsConfigFileName
+      );
+      if (tree.exists(runtimeTsConfigPath)) {
+        referencePath = runtimeTsConfigPath;
+      } else {
+        // Check for other possible runtime tsconfig file names
+        // TODO(leo): should we check if there are more than one runtime tsconfig files and throw an error?
+        for (const possibleRuntimeTsConfigFileName of possibleRuntimeTsConfigFileNames ??
+          []) {
+          const possibleRuntimeTsConfigPath = joinPathFragments(
+            dep.data.root,
+            possibleRuntimeTsConfigFileName
+          );
+          if (tree.exists(possibleRuntimeTsConfigPath)) {
+            referencePath = possibleRuntimeTsConfigPath;
+            break;
+          }
+        }
+      }
+    }
+    const relativePathToTargetRoot = relative(projectRoot, referencePath);
+    if (!newReferencesSet.has(relativePathToTargetRoot)) {
+      newReferencesSet.add(relativePathToTargetRoot);
+      // Make sure we unshift rather than push so that dependencies are built in the right order by TypeScript when it is run directly from the root of the workspace
+      references.unshift({ path: relativePathToTargetRoot });
+    }
+    if (!originalReferencesSet.has(relativePathToTargetRoot)) {
+      hasChanges = true;
+    }
+  }
+
+  hasChanges ||= newReferencesSet.size !== originalReferencesSet.size;
+
+  if (hasChanges) {
+    tsConfig.references = references;
+    writeJson(tree, tsConfigPath, tsConfig);
+  }
+
+  return hasChanges;
+}
+
+// Collect the dependencies of a project recursively sorted from root to leaf
+function collectProjectDependencies(
```

Comment:
Check if we have this logic elsewhere.

Reply:
Not in the way is used here and this one might get simplified anyways if it's confirmed we don't need transitive deps.